clusterctl/clusterdeployer/clusterdeployer_test.go

```diff
@@ -484,6 +483,72 @@ func TestCreateProviderComponentsScenarios(t *testing.T) {
 	}
 }
 
+func TestExtractMasterMachine(t *testing.T) {
+	const singleMasterName = "test-master"
+	multpleMasterNames := []string{"test-master-1", "test-master-2"}
+	const singleNodeName = "test-node"
+	multipleNodeNames := []string{"test-node-1", "test-node-2", "test-node-3"}
+
+	testCases := []struct {
+		name            string
+		inputMachines   []*clusterv1.Machine
+		expectedMasters *clusterv1.Machine
+		expectedNodes   []*clusterv1.Machine
+		expectedError   error
+	}{
+		{
+			name:            "success_1_master_1_node",
+			inputMachines:   generateMachines(),
+			expectedMasters: generateMasterNode(singleMasterName),
+			expectedNodes:   generateTestNodeMachines([]string{singleNodeName}),
+			expectedError:   nil,
+		},
+		{
+			name:            "success_1_master_multiple_nodes",
+			inputMachines:   generateValidExtractMasterMachineInput(singleMasterName, multipleNodeNames),
+			expectedMasters: generateMasterNode("test-master"),
+			expectedNodes:   generateTestNodeMachines([]string{"test-node-1", "test-node-2", "test-node-3"}),
+			expectedError:   nil,
+		},
+		{
+			name:            "fail_more_than_1_master_not_allowed",
+			inputMachines:   generateInvalidExtractMasterMachine(multpleMasterNames, multipleNodeNames),
+			expectedMasters: nil,
+			expectedNodes:   nil,
+			expectedError:   fmt.Errorf("expected one master, got: 2"),
+		},
+		{
+			name:            "fail_0_master_not_allowed",
+			inputMachines:   generateTestNodeMachines(multipleNodeNames),
+			expectedMasters: nil,
+			expectedNodes:   nil,
+			expectedError:   fmt.Errorf("expected one master, got: 0"),
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			actualMasters, actualNodes, actualError := extractMasterMachine(tc.inputMachines)
+
+			if tc.expectedError == nil && actualError != nil {
+				t.Fatalf("Expected error=[nil], Actual error=[%v]", actualError)
```

Comment:
Your test strings are backwards according to the go style (see https://github.com/golang/go/wiki/CodeReviewComments#useful-test-failures)

Something like `t.Fatalf("extractMasterMachine(%q) = %q; want %q", tc.inputMachines, actualError, tc.expectedError)` would be more along the lines of the style recommendation.

Reply:
Thank you for that link!